extensions/notebook/src/protocol/notebookUriHandler.ts

```diff
@@ -35,13 +34,13 @@ export class NotebookUriHandler implements vscode.UriHandler {
 	}
 
 	private open(uri: vscode.Uri): Promise<void> {
-		const data = querystring.parse(uri.query);
+		const data = uri.query.toString();
```

Comment:
Just curious, why do we need these changes to the open method?

Reply:
So the one issue I found when passing in URIs with `?x=1&y=2` is that it will cause the data variable to be corrupted and lose the query part of the URI. This was a way where we keep URI as a string entirely and not as a querystring.

Reply:
@lucyzhang929. Here is a quick comparison. The use of querystring is not able to handle the parameters query format that gets added to the end of URI for parameterization. Instead, `oldData` (shown below) gets set as a dictionary object which would then fail the openNotebook decoding of the URL and not get the proper parameters.

![image](https://user-images.githubusercontent.com/23587151/103311694-e42acc00-49cf-11eb-89f6-711fbfe9b795.png)
![image](https://user-images.githubusercontent.com/23587151/103311699-e725bc80-49cf-11eb-829e-20f15449a4b1.png)

Reply:
Can you give an example of the full URI passed in to `handleUri`?

Because it should look something like this : `azuredatastudio://microsoft.notebook/open?url=https://raw.githubusercontent.com/microsoft/sqlworkshops/master/sql2019lab/01_IntelligentPerformance/iqp/iqp_tablevariabledeferred.ipynb&x=1&y=2`

Note that x and y are values in the query string along with the URL - they are NOT query parameters on the notebook URL.

If done like this then I would expect the current approach to work fine.

Although if you have an actual reason for wanting to embed the query paramters into the URL itself then what should also work is to parse that URL separately - something like this : 

`const notebookUri = vscode.Uri.parse(uri.query.toString());`

and then pass that in and use notebookUri.query to access the query parameters at that point.

Reply:
Sure, so I do give an example of full URI actually in the output shown. This is shown in the gif as well.

`azuredatastudio://microsoft.notebook/open?https://raw.githubusercontent.com/VasuBhog/AzureDataStudio-Notebooks/main/Demo_Parameterization/Input.ipynb?x=1&y=2`

The reason to embed the query into the URI is so that when we load the new notebook in NotebookModel, we are able to utilize the URI to get the parameters notebookUri.query. I discussed this with @chlafreniere before the holidays and this was a solid route that enables both the proper setup of the URI for more generic use later in the future as well. 

We absolutely can not use the querystring method. The second part of what you said does not make sense. 
`const notebookUri = vscode.Uri.parse(uri.query.toString());` this would mean we setting the uri only to the query parameters. The above line is specifically ensuring we are getting the URI properly. If you look at the next function after `openNotebook` you will see we do validate this Uri where we actually set the Uri similar to how you have but in the next function.

Reply:
Ah I overlooked the URL part thanks for pointing that out. I will be sure to test out my work with this and adjust it to ensure this version works.

Reply:
> The query parameter string should be a set of key/value pairs. Your new code though is changing this so the top level query string is another full URL itself. This is not only meaning that previous links which used the old url= format are breaking but is also not following the standard URI format.

@Charles-Gagnon I believe that the data variable (set to uri.query.toString) can not be key/value pairs for the specific reason I showed in my quick comparison above (oldData vs data). If you see the way oldData is established it uses the key and value pairs of the URI, if I were to change it back to querystring and have the appropriate standard URI format (uri.query.toString for data) then the **parameters** that is injected will be misformatted and then we will have another error regarding opening the URL and it will throw the error `Failed to open URI`.

Moreover, `const notebookUri = vscode.Uri.parse(uri.query.toString());` gives the Error: _[UriError]: Scheme contains illegal characters._ since uri.query is `url=https://raw.githubusercontent.com/VasuBhog/AzureDataStudio-Notebooks/main/Demo_Parameterization/Input.ipynb`.

I think I would be able to set data like this - 
`const data = uri.query.toString();` which would then store "url=https://raw.githubusercontent.com/VasuBhog/AzureDataStudio-Notebooks/main/Demo_Parameterization/Input.ipynb" and then I can parse for everything after the url=. What are your thoughts on this?

Reply:
So yeah - my example wouldn't work directly since I was basing it off of your changes. 

Here's what I'm envisioning this will look like : 

1. You keep the parsing logic as it is 

`const data = querystring.parse(uri.query);`

This will end up with an object that has a single url property containing the full URL (including parameters) - it should be something like this : 

`{ url : 'https://raw.githubusercontent.com/VasuBhog/AzureDataStudio-Notebooks/main/Demo_Parameterization/Input.ipynb?x=1&y=2' }`

Which you'll then pass to openNotebook as was being done before.

But as I mentioned earlier I'm pretty sure you're going to have to encode the parameters here since you don't want them messing with the top level URI. So at some point you're likely going to have to decode that as well, which you probably want to do here. Something like `decodeURIComponent` should do the trick.

Reply:
It actually looks like we already decode the URL on line 51 so you probably don't even need to do anything else for that.

Reply:
> `const data = querystring.parse(uri.query);`

So I have the exact example of how this will look with parameters query (discussed after Lucy's comment) added to URI ie '...Input.ipynb?x=1&y=2'. As I said we can not use `querystring.parse` shown below:

![image](https://user-images.githubusercontent.com/23587151/103715160-bcf67080-4f74-11eb-962b-760886750688.png)
![image](https://user-images.githubusercontent.com/23587151/103715181-cda6e680-4f74-11eb-9b69-dfbd0a76b896.png)

As you can see the querystring version would make key and value pairs for the **parameters**. I think having my version of this data is generic and works for the standard URL format. 

This will then be decoded properly and we are able to ensure the parameters query for this URI (?x=1&y=2) is set to the URI.

Reply:
That's why you need to have the caller encode the URL - to avoid the parameters being seen as part of the top level URI (the azuredatastudio:// part with the url parameter). So the full URL would be something like this : 

`azuredatastudio://microsoft.notebook/open?url=https%3A%2F%2Fraw.githubusercontent.com%2FVasuBhog%2FAzureDataStudio-Notebooks%2Fmain%2FDemo_Parameterization%2FInput.ipynb%3Fx%3D1%26y%3D2`

You then pull out the url value : `https%3A%2F%2Fraw.githubusercontent.com%2FVasuBhog%2FAzureDataStudio-Notebooks%2Fmain%2FDemo_Parameterization%2FInput.ipynb%3Fx%3D1%26y%3D2` and pass it through decodeURIComponent to get the actual URL to open : `https://raw.githubusercontent.com/VasuBhog/AzureDataStudio-Notebooks/main/Demo_Parameterization/Input.ipynb?x=1&y=2` which can then be parsed as a normal URI to get the query parameters.

Reply:
After discussions with @Charles-Gagnon, it seems that the encoded URI with parameters is already decoded from the extHostUrls (VSCode side). As a result they are decoded by the time we get the uri.query back and set as a string. 

For now I will push my update using the split method, but we will discuss further with @chlafreniere.